File: benchmarks/references/ground-truth/Python/tc_020_orr_by_subgroup.py
#!/usr/bin/env python3
"""TC-020 Ground Truth: ORR (Objective Response Rate) by Subgroup (Python)

Part of the ASA Biopharm AI/ML WG Agentic AI Benchmark.

Computes ORR (CR+PR rate) by pre-specified subgroups with:
  - Subgroup-level ORR (n/N, %) for each arm
  - Cochran-Mantel-Haenszel chi-square test for treatment-by-subgroup interaction
  - Forest-plot-ready output: ORR difference (Exp - Ctrl) with 95% CI per subgroup

Dependencies: pandas, numpy, scipy
Usage: python tc_020_orr_by_subgroup.py --seed 42 --n 200 --output results.json
"""

import argparse
import json
from pathlib import Path
import numpy as np
import pandas as pd
from scipy import stats


def generate_tumor_response(seed=42, n_subjects=200):
    """Generate synthetic tumor response dataset."""
    rng = np.random.default_rng(seed)

    trt = rng.choice([0, 1], size=n_subjects)
    sex = rng.choice(["Male", "Female"], size=n_subjects, p=[0.55, 0.45])
    age = np.round(rng.normal(58, 12, n_subjects)).astype(int)
    agegr1 = np.where(age < 65, "<65", ">=65")
    ecog = rng.choice([0, 1], size=n_subjects, p=[0.6, 0.4])

    # Response probability
    orr_prob = np.where(trt == 1,
                        np.where(ecog == 0, 0.45, 0.30),
                        np.where(ecog == 0, 0.25, 0.15))

    is_responder = rng.binomial(1, orr_prob)
    is_cr = np.where(is_responder == 1, rng.binomial(1, 0.3, n_subjects), 0)

    bor = np.where(is_cr == 1, "CR",
                   np.where(is_responder == 1, "PR",
                            rng.choice(["SD", "PD"], size=n_subjects, p=[0.4, 0.6])))

    aval = np.where(np.isin(bor, ["CR", "PR"]), 1, 0)

    return pd.DataFrame({
        "USUBJID": [f"SUBJ-{i+1:04d}" for i in range(n_subjects)],
        "STUDYID": "BENCHMARK-001",
        "TRT01PN": trt,
        "TRT01A": np.where(trt == 0, "Control", "Experimental"),
        "SEX": sex,
        "AGEGR1": agegr1,
        "ECOG": ecog,
        "BOR": bor,
        "AVAL": aval,
    })


def wilson_ci(x, n, conf=0.95):
    """Wilson score confidence interval for a proportion."""
    z = stats.norm.ppf(1 - (1 - conf) / 2)
    p = x / n
    denom = 1 + z**2 / n
    center = (p + z**2 / (2 * n)) / denom
    margin = z * np.sqrt(p * (1 - p) / n + z**2 / (4 * n**2)) / denom
    return center - margin, center + margin


def compute_orr_by_subgroup(data):
    """Compute ORR by subgroup with interaction tests."""
    subgroups = ["SEX", "AGEGR1", "ECOG"]
    subgroup_levels = {
        "SEX": ["Male", "Female"],
        "AGEGR1": ["<65", ">=65"],
        "ECOG": ["0", "1"],
    }

    overall_exp = data[data["TRT01PN"] == 1]
    overall_ctrl = data[data["TRT01PN"] == 0]

    orr_exp = overall_exp["AVAL"].sum() / len(overall_exp) * 100
    orr_ctrl = overall_ctrl["AVAL"].sum() / len(overall_ctrl) * 100

    ci_exp = wilson_ci(overall_exp["AVAL"].sum(), len(overall_exp))
    ci_ctrl = wilson_ci(overall_ctrl["AVAL"].sum(), len(overall_ctrl))

    overall = {
        "orr_experimental": round(orr_exp, 1),
        "orr_control": round(orr_ctrl, 1),
        "n_experimental": len(overall_exp),
        "n_control": len(overall_ctrl),
        "responders_experimental": int(overall_exp["AVAL"].sum()),
        "responders_control": int(overall_ctrl["AVAL"].sum()),
        "ci_lower_experimental": round(ci_exp[0] * 100, 1),
        "ci_upper_experimental": round(ci_exp[1] * 100, 1),
        "ci_lower_control": round(ci_ctrl[0] * 100, 1),
        "ci_upper_control": round(ci_ctrl[1] * 100, 1),
        "orr_difference": round(orr_exp - orr_ctrl, 1),
    }

    subgroup_results = []
    for sg in subgroups:
        for lvl in subgroup_levels[sg]:
            sg_exp = data[(data["TRT01PN"] == 1) & (data[sg] == lvl)]
            sg_ctrl = data[(data["TRT01PN"] == 0) & (data[sg] == lvl)]

            if len(sg_exp) == 0 or len(sg_ctrl) == 0:
                continue

            orr_e = sg_exp["AVAL"].sum() / len(sg_exp) * 100
            orr_c = sg_ctrl["AVAL"].sum() / len(sg_ctrl) * 100
            ci_e = wilson_ci(sg_exp["AVAL"].sum(), len(sg_exp))
            ci_c = wilson_ci(sg_ctrl["AVAL"].sum(), len(sg_ctrl))

            diff = orr_e - orr_c
            se_diff = np.sqrt(
                (orr_e / 100 * (1 - orr_e / 100)) / len(sg_exp) +
                (orr_c / 100 * (1 - orr_c / 100)) / len(sg_ctrl)
            ) * 100
            z = stats.norm.ppf(0.975)

            subgroup_results.append({
                "subgroup": sg,
                "level": lvl,
                "orr_experimental": round(orr_e, 1),
                "orr_control": round(orr_c, 1),
                "n_experimental": len(sg_exp),
                "n_control": len(sg_ctrl),
                "responders_experimental": int(sg_exp["AVAL"].sum()),
                "responders_control": int(sg_ctrl["AVAL"].sum()),
                "ci_lower_experimental": round(ci_e[0] * 100, 1),
                "ci_upper_experimental": round(ci_e[1] * 100, 1),
                "ci_lower_control": round(ci_c[0] * 100, 1),
                "ci_upper_control": round(ci_c[1] * 100, 1),
                "orr_difference": round(diff, 1),
                "diff_ci_lower": round(diff - z * se_diff, 1),
                "diff_ci_upper": round(diff + z * se_diff, 1),
            })

    # CMH interaction tests
    interaction_pvalues = []
    for sg in subgroups:
        # Build 2x2xK contingency table
        levels = subgroup_levels[sg]
        # Use scipy.stats.chi2_contingency on the 2x2 table pooled
        # For CMH, use stratified analysis
        try:
            from statsmodels.stats.contingency_tables import StratifiedTable
            strata = []
            for lvl in levels:
                d = data[data[sg] == lvl]
                if len(d) == 0:
                    continue
                tbl = pd.crosstab(d["TRT01PN"], d["AVAL"])
                if tbl.shape == (2, 2):
                    strata.append(tbl.values)

            if len(strata) >= 2:
                st = StratifiedTable(strata)
                cmh_p = st.test_null_odds().pvalue
                common_or = st.oddsratio_pooled
                interaction_pvalues.append({
                    "subgroup": sg,
                    "cmh_p_value": round(float(cmh_p), 4),
                    "cmh_common_or": round(float(common_or), 4),
                })
        except ImportError:
            # statsmodels not available; use scipy CMH approximation
            # Fall back to simple chi-square on treatment x response
            # stratified by subgroup
            for lvl in levels:
                d = data[data[sg] == lvl]
                if len(d) == 0:
                    continue
                tbl = pd.crosstab(d["TRT01PN"], d["AVAL"])
                if tbl.shape == (2, 2):
                    chi2, p, _, _ = stats.chi2_contingency(tbl.values)
                    interaction_pvalues.append({
                        "subgroup": f"{sg}_{lvl}",
                        "cmh_p_value": round(float(p), 4),
                        "cmh_common_or": None,
                    })

    return {
        "test_case_id": "TC-020",
        "variant_id": None,
        "language": "Python",
        "package": "scipy",
        "package_version": stats.__version__ if hasattr(stats, "__version__") else "unknown",
        "overall": overall,
        "subgroups": subgroup_results,
        "interaction_pvalues": interaction_pvalues,
        "seed": None,
    }


def main():
    parser = argparse.ArgumentParser(
        description="TC-020: ORR by Subgroup (Python)"
    )
    parser.add_argument("--seed", type=int, default=42, help="Random seed")
    parser.add_argument("--n", type=int, default=200, help="Number of subjects")
    parser.add_argument("--data", type=str, default=None,
                        help="Shared tumor response CSV")
    parser.add_argument("--output", type=str, default=None, help="Output JSON path")
    args = parser.parse_args()

    print(f"TC-020: ORR by Subgroup (Python) — seed={args.seed}, n={args.n}")

    if args.data:
        data = pd.read_csv(args.data)
        print(f"Loaded shared data with {len(data)} subjects")
    else:
        data = generate_tumor_response(seed=args.seed, n_subjects=args.n)
        print(f"Generated tumor response data with {len(data)} subjects")

    result = compute_orr_by_subgroup(data)
    result["seed"] = args.seed
    result["variant_id"] = f"v{args.seed}"

    print("\n" + "─" * 50)
    o = result["overall"]
    print(f"Overall ORR:  Exp={o['orr_experimental']}%, "
          f"Ctrl={o['orr_control']}%, Diff={o['orr_difference']}%")
    print(f"Subgroups analyzed: {len(result['subgroups'])}")
    print(f"Interaction tests: {len(result['interaction_pvalues'])}")
    print("─" * 50)

    print("\n=== BENCHMARK OUTPUT ===")
    print(json.dumps(result, indent=2, default=str))
    print("=== END OUTPUT ===")

    if args.output:
        outpath = Path(args.output)
        outpath.parent.mkdir(parents=True, exist_ok=True)
        with open(outpath, "w") as f:
            json.dump(result, f, indent=2, default=str)
        print(f"Wrote output to: {outpath}")


if __name__ == "__main__":
    main()

File: benchmarks/references/ground-truth/R/tc-020-orr-by-subgroup.R
#!/usr/bin/env Rscript
# TC-020 Ground Truth: ORR (Objective Response Rate) by Subgroup
# Part of the ASA Biopharm AI/ML WG Agentic AI Benchmark
#
# Computes ORR (CR+PR rate) by pre-specified subgroups with:
#   - Subgroup-level ORR (n/N, %) for each arm
#   - Cochran-Mantel-Haenszel chi-square test for treatment-by-subgroup interaction
#   - Forest-plot-ready output: ORR difference (Exp - Ctrl) with 95% CI per subgroup
#
# Dependencies: dplyr, jsonlite, stats
# Usage: Rscript tc-020-orr-by-subgroup.R --seed 42 --n 200 --output results.json

library(dplyr)
library(jsonlite)
library(stats)

# Source common utilities
source("common/data-generation.R")

# ─────────────────────────────────────────────────────
# Command-line argument parsing
# ─────────────────────────────────────────────────────

parse_args <- function() {
  args <- commandArgs(trailingOnly = TRUE)
  get_arg <- function(flag, default) {
    idx <- which(args == flag)
    if (length(idx) > 0 && idx < length(args)) return(args[idx + 1])
    return(default)
  }
  list(
    seed = as.integer(get_arg("--seed", "42")),
    n = as.integer(get_arg("--n", "200")),
    data = get_arg("--data", NA),
    output = get_arg("--output", NA)
  )
}

# ─────────────────────────────────────────────────────
# Synthetic Tumor Response Dataset Generation
# ─────────────────────────────────────────────────────

generate_tumor_response <- function(seed = 42, n_subjects = 200) {
  set.seed(seed)

  trt <- sample(0:1, n_subjects, replace = TRUE)

  # Subgroups: SEX (Male/Female), AGEGR1 (<65/>=65), ECOG (0/1)
  sex <- sample(c("Male", "Female"), n_subjects, replace = TRUE, prob = c(0.55, 0.45))
  age <- round(rnorm(n_subjects, mean = 58, sd = 12))
  agegr1 <- ifelse(age < 65, "<65", ">=65")
  ecog <- sample(0:1, n_subjects, replace = TRUE, prob = c(0.6, 0.4))

  # Response probability depends on treatment and ECOG
  # Experimental: ORR ~45% (ECOG 0), ~30% (ECOG 1)
  # Control: ORR ~25% (ECOG 0), ~15% (ECOG 1)
  orr_prob <- ifelse(trt == 1,
                     ifelse(ecog == 0, 0.45, 0.30),
                     ifelse(ecog == 0, 0.25, 0.15))

  # Best overall response: CR, PR, SD, PD
  # Simplified: response = rbinom, then split CR vs PR
  is_responder <- rbinom(n_subjects, 1, orr_prob)
  # Among responders: ~30% CR, 70% PR
  is_cr <- ifelse(is_responder == 1, rbinom(n_subjects, 1, 0.3), 0)
  bor <- ifelse(is_cr == 1, "CR",
                ifelse(is_responder == 1, "PR",
                       sample(c("SD", "PD"), n_subjects, replace = TRUE, prob = c(0.4, 0.6))))

  data.frame(
    USUBJID = sprintf("SUBJ-%04d", 1:n_subjects),
    STUDYID = "BENCHMARK-001",
    TRT01PN = trt,
    TRT01A = ifelse(trt == 0, "Control", "Experimental"),
    SEX = sex,
    AGEGR1 = agegr1,
    ECOG = ecog,
    BOR = bor,
    AVAL = ifelse(bor %in% c("CR", "PR"), 1, 0),
    stringsAsFactors = FALSE
  )
}

# ─────────────────────────────────────────────────────
# TC-020 Core Computation
# ─────────────────────────────────────────────────────

compute_orr_by_subgroup <- function(data) {
  subgroups <- c("SEX", "AGEGR1", "ECOG")
  subgroup_labels <- list(
    SEX = c("Male", "Female"),
    AGEGR1 = c("<65", ">=65"),
    ECOG = c("0", "1")
  )

  # Overall ORR
  overall_exp <- data[data$TRT01PN == 1, ]
  overall_ctrl <- data[data$TRT01PN == 0, ]

  overall_orr_exp <- sum(overall_exp$AVAL) / nrow(overall_exp) * 100
  overall_orr_ctrl <- sum(overall_ctrl$AVAL) / nrow(overall_ctrl) * 100

  # Overall CI (Wilson interval)
  wilson_ci <- function(x, n, conf = 0.95) {
    z <- qnorm(1 - (1 - conf) / 2)
    p <- x / n
    denom <- 1 + z^2 / n
    center <- (p + z^2 / (2 * n)) / denom
    margin <- z * sqrt(p * (1 - p) / n + z^2 / (4 * n^2)) / denom
    c(lower = center - margin, upper = center + margin)
  }

  ci_exp <- wilson_ci(sum(overall_exp$AVAL), nrow(overall_exp))
  ci_ctrl <- wilson_ci(sum(overall_ctrl$AVAL), nrow(overall_ctrl))

  overall <- list(
    orr_experimental = round(overall_orr_exp, 1),
    orr_control = round(overall_orr_ctrl, 1),
    n_experimental = nrow(overall_exp),
    n_control = nrow(overall_ctrl),
    responders_experimental = sum(overall_exp$AVAL),
    responders_control = sum(overall_ctrl$AVAL),
    ci_lower_experimental = round(ci_exp["lower"] * 100, 1),
    ci_upper_experimental = round(ci_exp["upper"] * 100, 1),
    ci_lower_control = round(ci_ctrl["lower"] * 100, 1),
    ci_upper_control = round(ci_ctrl["upper"] * 100, 1),
    orr_difference = round(overall_orr_exp - overall_orr_ctrl, 1)
  )

  # Subgroup-level
  subgroup_results <- list()
  for (sg in subgroups) {
    levels <- subgroup_labels[[sg]]
    for (lvl in levels) {
      sg_exp <- data[data$TRT01PN == 1 & data[[sg]] == lvl, ]
      sg_ctrl <- data[data$TRT01PN == 0 & data[[sg]] == lvl, ]

      if (nrow(sg_exp) == 0 || nrow(sg_ctrl) == 0) next

      orr_e <- sum(sg_exp$AVAL) / nrow(sg_exp) * 100
      orr_c <- sum(sg_ctrl$AVAL) / nrow(sg_ctrl) * 100
      ci_e <- wilson_ci(sum(sg_exp$AVAL), nrow(sg_exp))
      ci_c <- wilson_ci(sum(sg_ctrl$AVAL), nrow(sg_ctrl))

      # Risk difference CI (normal approximation)
      diff <- orr_e - orr_c
      se_diff <- sqrt((orr_e / 100 * (1 - orr_e / 100)) / nrow(sg_exp) +
                       (orr_c / 100 * (1 - orr_c / 100)) / nrow(sg_ctrl)) * 100
      z <- qnorm(0.975)
      diff_ci_lower <- round(diff - z * se_diff, 1)
      diff_ci_upper <- round(diff + z * se_diff, 1)

      subgroup_results <- c(subgroup_results, list(list(
        subgroup = sg,
        level = lvl,
        orr_experimental = round(orr_e, 1),
        orr_control = round(orr_c, 1),
        n_experimental = nrow(sg_exp),
        n_control = nrow(sg_ctrl),
        responders_experimental = sum(sg_exp$AVAL),
        responders_control = sum(sg_ctrl$AVAL),
        ci_lower_experimental = round(ci_e["lower"] * 100, 1),
        ci_upper_experimental = round(ci_e["upper"] * 100, 1),
        ci_lower_control = round(ci_c["lower"] * 100, 1),
        ci_upper_control = round(ci_c["upper"] * 100, 1),
        orr_difference = round(diff, 1),
        diff_ci_lower = diff_ci_lower,
        diff_ci_upper = diff_ci_upper
      )))
    }
  }

  # Interaction p-value (CMH test for each subgroup)
  interaction_pvalues <- list()
  for (sg in subgroups) {
    # Create 2x2x2 table: treatment x response x subgroup
    tbl <- table(data$TRT01PN, data$AVAL, data[[sg]])
    if (all(dim(tbl) >= c(2, 2, 2))) {
      cmh <- mantelhaen.test(tbl)
      interaction_pvalues <- c(interaction_pvalues, list(list(
        subgroup = sg,
        cmh_p_value = round(cmh$p.value, 4),
        cmh_common_or = round(as.numeric(cmh$estimate), 4)
      )))
    }
  }

  list(
    test_case_id = "TC-020",
    variant_id = NA,
    language = "R",
    package = "stats",
    package_version = R.version$major,
    overall = overall,
    subgroups = subgroup_results,
    interaction_pvalues = interaction_pvalues,
    seed = NA
  )
}

# ─────────────────────────────────────────────────────
# Main
# ─────────────────────────────────────────────────────

if (sys.nframe() == 0) {
  opts <- parse_args()
  seed <- opts$seed
  n <- opts$n

  cat(sprintf("TC-020: ORR by Subgroup (R) — seed=%d, n=%d\\n", seed, n))

  if (!is.na(opts$data)) {
    data <- read_shared_data(opts$data)
    cat(sprintf("Loaded shared tumor response data with %d subjects\\n", nrow(data)))
  } else {
    data <- generate_tumor_response(seed = seed, n_subjects = n)
    cat(sprintf("Generated tumor response data with %d subjects\\n", nrow(data)))
  }

  result <- compute_orr_by_subgroup(data)
  result$seed <- seed
  result$variant_id <- paste0("v", seed)

  cat("\\n──────────────────────────────────────────────\\n")
  cat(sprintf("Overall ORR:  Exp=%.1f%%, Ctrl=%.1f%%, Diff=%.1f%%\\n",
              result$overall$orr_experimental, result$overall$orr_control,
              result$overall$orr_difference))
  cat(sprintf("Subgroups analyzed: %d\\n", length(result$subgroups)))
  cat(sprintf("Interaction tests: %d\\n", length(result$interaction_pvalues)))
  cat("──────────────────────────────────────────────\\n")

  print_output(result)

  if (!is.na(opts$output)) {
    write_output(result, opts$output)
  }
}
